From a96251835a47259ec51f5d4922334816c8185fdc Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 4 Jan 2017 09:51:02 +0800
Subject: [PATCH 1454/5242] MLK-13715-2 video: mxsfb: add global alpha mode
 for overlay fb

commit  8e42bfeac8528a6abda9fe30fdd8faa60d1d4785 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add global alpha mode for overlay framebuffer when the
overlay fb has no alpha channel, which means set a fix
alpha value which is used during the alpha blending
between AS and framebuffer.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index 359ace0..481722b 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -219,6 +219,7 @@ struct mxsfb_layer {
 	int			registered;
 	uint32_t		usage;
 	int			blank_state;
+	uint32_t 		global_alpha;
 
 	struct mxsfb_layer_ops	*ops;
 
@@ -1587,7 +1588,7 @@ static void overlayfb_disable(struct mxsfb_layer *ofb)
 static void overlayfb_setup(struct mxsfb_layer *ofb)
 {
 	uint32_t as_next_buf, as_ctrl = 0;
-	uint8_t format, alpha_ctrl;
+	uint8_t format, alpha_ctrl, global_alpha_en = 0;
 	struct mxsfb_info *fbi = ofb->fbi;
 	struct fb_var_screeninfo *var = &ofb->ol_fb->var;
 
@@ -1595,11 +1596,15 @@ static void overlayfb_setup(struct mxsfb_layer *ofb)
 	as_next_buf = ofb->video_mem_phys;
 	writel(as_next_buf, fbi->base + LCDC_AS_NEXT_BUF);
 
+	/* clear the LCDC_AS_CTRL */
+	writel(0x0, fbi->base + LCDC_AS_CTRL);
+
 	switch (var->grayscale) {
 	case 0: /* color */
 		switch (var->bits_per_pixel) {
 		case 16: /* RGB565 */
 			format = 0xE;
+			global_alpha_en = 1;
 			break;
 		case 32: /* ARGB8888 */
 			format = 0x0;
@@ -1617,6 +1622,7 @@ static void overlayfb_setup(struct mxsfb_layer *ofb)
 			break;
 		case V4L2_PIX_FMT_XRGB32:
 			format = 0x4;
+			global_alpha_en = 1;
 			break;
 		case V4L2_PIX_FMT_ARGB555:
 			format = 0x8;
@@ -1626,12 +1632,15 @@ static void overlayfb_setup(struct mxsfb_layer *ofb)
 			break;
 		case V4L2_PIX_FMT_RGB555:
 			format = 0xC;
+			global_alpha_en = 1;
 			break;
 		case V4L2_PIX_FMT_RGB444:
 			format = 0xD;
+			global_alpha_en = 1;
 			break;
 		case V4L2_PIX_FMT_RGB565:
 			format = 0xE;
+			global_alpha_en = 1;
 			break;
 		default:
 			return;
@@ -1640,9 +1649,11 @@ static void overlayfb_setup(struct mxsfb_layer *ofb)
 	}
 	as_ctrl |= ((format & 0xf) << 4);
 
-	/*XXX Only support embbeded alpha mode first */
-	alpha_ctrl = ALPHA_CTRL_EMBEDDED;
+	alpha_ctrl = global_alpha_en ? ALPHA_CTRL_OVERRIDE :
+				       ALPHA_CTRL_EMBEDDED;
 	as_ctrl |= ((alpha_ctrl & 0x3) << 1);
+	if (global_alpha_en)
+		as_ctrl |= ((ofb->global_alpha & 0xff) << 8);
 
 	writel(as_ctrl, fbi->base + LCDC_AS_CTRL);
 }
@@ -1893,6 +1904,7 @@ static void init_mxsfb_overlay(struct mxsfb_info *fbi,
 	ofb->ops = &ofb_ops;
 	ofb->usage = 0;
 	ofb->blank_state = -1;
+	ofb->global_alpha = 255;
 	ofb->fbi = fbi;
 }
 
-- 
1.7.9.5

